src/ray/core_worker/core_worker.cc

```diff
@@ -4454,45 +4454,30 @@ void CoreWorker::CancelActorTaskOnExecutor(WorkerID caller_worker_id,
                  caller_worker_id,
                  on_canceled = std::move(on_canceled),
                  is_async_actor]() {
-    bool is_task_running = false;
-    TaskSpecification spec;
-    RayFunction func;
-    std::string concurrency_group_name;
-
-    bool is_task_queued_or_executing =
-        task_receiver_->CancelQueuedActorTask(caller_worker_id, task_id);
-
-    // If a task is already running, we send a cancel request.
-    // Right now, we can only cancel async actor tasks.
-    if (is_task_queued_or_executing) {
+    // If the task was still queued (not running yet), `CancelQueuedActorTask` will cancel
+    // it. If it is already running, we attempt to cancel it.
+    bool success = false;
+    bool is_running = false;
+    bool task_present = task_receiver_->CancelQueuedActorTask(caller_worker_id, task_id);
+    if (task_present) {
       {
         absl::MutexLock lock(&mutex_);
-        auto it = current_tasks_.find(task_id);
-        is_task_running = it != current_tasks_.end();
-        if (is_task_running) {
-          spec = it->second;
-          func = RayFunction(spec.GetLanguage(), spec.FunctionDescriptor());
-          concurrency_group_name = spec.ConcurrencyGroupName();
-        }
+        is_running = running_tasks_.find(task_id) != running_tasks_.end();
       }
 
-      if (is_task_running && is_async_actor) {
-        options_.cancel_async_task(task_id, func, concurrency_group_name);
+      // Attempt to cancel the task if it's running.
```

Comment:
Should L4468 - L4476 move to the same lock scope as `is_running = ...` above?

What happens if `is_running` is false, but the task is added to `running_tasks_` between the assignment to `is_running` and the `if (is_running && …)` check? It looks like success would be set to true, even though the task is still running. Is this possible to happen?

Reply:
I think you're right, but I need to think more closely about this. I don't think we should hold the core worker mutex here because it could cause deadlock: we need to grab the GIL in order to execute the python code to cancel the running task. If there is a call that holds the GIL that calls into the core worker, that could deadlock.

Reply:
You're right, but it's a separate issue from the one this PR fixes, so I'm going to proceed with merging and take this as a follow up (that I'll make sure we do)